src/Dfc.CourseDirectory.WebV2/Features/Providers/EditProviderType.cs

```diff
@@ -40,25 +52,77 @@ public async Task<Command> Handle(Query request, CancellationToken cancellationT
                 ProviderId = request.ProviderId
             });
 
+            var tLevelDefinitions = await _sqlQueryDispatcher.ExecuteQuery(new SqlQueries.GetTLevelDefinitions());
+            var providerTLevelDefinitions = provider.ProviderType.HasFlag(ProviderType.TLevels)
+                ? await _sqlQueryDispatcher.ExecuteQuery(new SqlQueries.GetTLevelDefinitionsForProvider { ProviderId = request.ProviderId })
+                : Enumerable.Empty<SqlModels.TLevelDefinition>();
+
             return new Command()
             {
                 ProviderId = provider.Id,
-                ProviderType = provider.ProviderType
+                ProviderType = provider.ProviderType,
+                ProviderTLevelDefinitions = ProviderTLevelDefinitions(tLevelDefinitions, providerTLevelDefinitions.Select(pd => pd.TLevelDefinitionId))
             };
         }
 
-        public async Task<Success> Handle(Command request, CancellationToken cancellationToken)
+        public async Task<OneOf<ModelWithErrors<Command>, Success>> Handle(Command request, CancellationToken cancellationToken)
         {
+            var tLevelDefinitions = await _sqlQueryDispatcher.ExecuteQuery(new SqlQueries.GetTLevelDefinitions());
+
+            var validationResult = await new CommandValidator(tLevelDefinitions).ValidateAsync(request);
+
+            if (!validationResult.IsValid)
+            {
+                request.ProviderTLevelDefinitions = ProviderTLevelDefinitions(tLevelDefinitions, request.SelectedProviderTLevelDefinitionIds);
+                return new ModelWithErrors<Command>(request, validationResult);
+            }
+
             await _cosmosDbQueryDispatcher.ExecuteQuery(new UpdateProviderType()
             {
                 ProviderId = request.ProviderId,
                 ProviderType = request.ProviderType
             });
 
+            await _sqlQueryDispatcher.ExecuteQuery(new SqlQueries.UpsertProviderTLevelDefinitions
```

Comment:
Isn't there a requirement to maintain the T-Level Definitions for a provider even when T-Levels is removed?

Reply:
Not in this story! (I've played with it and given the validation requirements this is the only behaviour that makes sense.)